testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/user/profile/DynamicUserProfileProvider.java

```diff
@@ -0,0 +1,213 @@
+        AmphibianProviderFactory.super.onCreate(session, realm, model);
+    }
+
+    @Override
+    public void onUpdate(KeycloakSession session, RealmModel realm, ComponentModel oldModel, ComponentModel newModel) {
+        AmphibianProviderFactory.super.onUpdate(session, realm, oldModel, newModel);
+        removeMetadataFromCache(realm);
+    }
+
+    @Override
+    public void preRemove(KeycloakSession session, RealmModel realm, ComponentModel model) {
+        AmphibianProviderFactory.super.preRemove(session, realm, model);
+        removeMetadataFromCache(realm);
+    }
+
+    @Override
+    public void init(Config.Scope config) {
+        super.init(config);
+        METADATA_CACHE = Collections.synchronizedMap(new LinkedHashMap<String, Map<UserProfileContext, UserProfileMetadata>>(100, 0.75f, true) {
+            @Override
+            protected boolean removeEldestEntry(Map.Entry eldest) {
+                return METADATA_CACHE.size() > 100; // make this configurable
+            }
+        });
+    }
+
+    private UserProfileMetadata decorateUserProfile(UserProfileMetadata metadata, KeycloakSession session) {
+        ComponentModel model = getComponentModel(session);
+
+        if (model == null) {
+            return metadata;
+        }
+
+        // need to clone otherwise changes to profile config are going to be reflected in the default config
+        UserProfileMetadata decoratedMetadata = metadata.clone();
+
+        String config = model.getConfig().getFirst("config");
+
+        try {
+            Object attribute = JsonSerialization.readValue(config, Map.class).get("validateConfigAttribute");
+
+            if (attribute != null && Boolean.parseBoolean(attribute.toString())) {
+                decoratedMetadata.addAttribute("validateConfigAttribute", createRequiredValidator("required_dynamic_attribute"));
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to parse config", e);
+        }
+
+        return decoratedMetadata;
+    }
+
+    private void removeMetadataFromCache(RealmModel realm) {
+        synchronized (METADATA_CACHE) {
+            METADATA_CACHE.remove(realm.getId());
+        }
+    }
+
+    private ComponentModel getComponentModel(KeycloakSession session) {
+        RealmModel realm = session.getContext().getRealm();
+        return realm.getComponentsStream(null, UserProfileProvider.class.getName())
+                .findAny()
+                .orElse(null);
+    }
+
+    private AttributeValidatorMetadata createRequiredValidator(String message) {
+        return new AttributeValidatorMetadata(message, context -> {
+            Map.Entry<String, List<String>> attribute = context.getAttribute();
+            List<String> values = attribute.getValue();
+            AttributeMetadata metadata = context.getMetadata();
+
+            if (!metadata.isOptional() && values.isEmpty()) {
```

Comment:
Is this validation enough? How are values with blank strings handled? Blank string should be treated as required validation violation.

Reply:
This is for testing. You should change that once you start the config work.

Reply:
OK, thanks